Call cat_ranges in blockcache for async filesystems

monken opened the pull request:
The CachingFileSystem can benefit from calling `cat_ranges` for asynchronous filesystems to concurrently download blocks. This speeds up filling the cache significantly for filesystems like HTTP and S3FS.

martindurant commented on fsspec/caching.py:
```diff
@@ -93,16 +114,30 @@ def _fetch(self, start, end):
+        if not ranges:
+            return self.cache[start:end]
+
+        if self.multi_fetcher:
```
Since .cat_ranges always exists and falls back to getting each block serially, should we just always call that?

monken replied:
In `caching.py`, I wanted to maintain backwards compatibility. If someone uses the `MMapCache` class directly, their code would break if we require the `multi_fetcher` style interface.

martindurant replied:
OK, but let's agree to eventually deprecate this

martindurant commented on fsspec/implementations/cached.py:
```diff
@@ -400,7 +400,21 @@ def _open(
+            )
+
+        multi_fetcher = (
+            None if not self.fs.async_impl or self.compression else _fetch_ranges
```
It might still be useful for a sync filesystem, since cat_ranges might combine some adjacent reads.

martindurant commented:
Sorry I let this one slide - I still think it's useful. I have just the one comment above.

martindurant commented:
Can you please solve the conflict, and I can have another look?

monken commented:
@martindurant sorry that I abandoned this for so long. I fixed the merge conflicts and updated the tests. Please let me know if you still seem this valuable.
